### PR TITLE
Postprocessing refactor img2img

### PR DESCRIPTION
**Pull request by yiyixuxu**
Refactor `StableDiffusionImg2ImgPipeline` with `VaeImageProcessor.postprocess`

based on the discussion in this PR https://github.com/huggingface/diffusers/pull/2943

Note:
I removed `#copy from` statement for below methods, will need to add it back after we refactor `StableDiffusionPipeline`
* `__init__`
* `run_safety_checker`
* `decode_latents`

I think I addressed all the new comments from @pcuenca and @williamberman, but let me know if I missed anything

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 1]
```diff
@@ -738,27 +750,19 @@ def __call__(
...
-            deprecate("Unsupported output_type", "1.0.0", deprecation_message, standard_warn=False)
-            output_type = "np"
-
-        if output_type == "latent":
+        if not output_type == "latent":
+            image = self.vae.decode(latents / self.vae.config.scaling_factor).sample
+            image, has_nsfw_concept = self.run_safety_checker(image, device, prompt_embeds.dtype)
+        else:
             image = latents
             has_nsfw_concept = None
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:758]: `has_nsfw_concept` is now either a list of bool or `None` (when didn't run safety_checker, i.e. `self.safety_checker is None` or `output_type=='latents'`)

I think this is consistent with the documentation as @pcuenca pointed out. And it is consistent with how it's used in current pipelines so I think it's ok

https://github.com/huggingface/diffusers/blob/7c1bb9ae81c97f6f10cf36784c17a51ebddeea08/src/diffusers/pipelines/stable_diffusion/__init__.py#L30-L32

File: tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py [hunk 2]
```diff
@@ -96,45 +101,32 @@ def get_dummy_components(self):
...
         device = "cpu"  # ensure determinism for the device-dependent torch.Generator
         components = self.get_dummy_components()
         sd_pipe = StableDiffusionImg2ImgPipeline(**components)
-        sd_pipe.image_processor = VaeImageProcessor(vae_scale_factor=sd_pipe.vae_scale_factor, do_normalize=False)
+        sd_pipe.image_processor = VaeImageProcessor(vae_scale_factor=sd_pipe.vae_scale_factor, do_normalize=True)
         sd_pipe = sd_pipe.to(device)
         sd_pipe.set_progress_bar_config(disable=None)
 
         inputs = self.get_dummy_inputs(device)
+        inputs["image"] = inputs["image"] / 2 + 0.5
```
Comment (yiyixuxu) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py:129]: The input `image` is a random pytorch tensor with value range `[0,1]` - since it is a pytorch tensor, it was not normalized in the previous `preprocess` method (see https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L74), however it was still denormalized in postprocess

Now with VaeImageProcessor, we will normalize the image in `preprocess` and denormalize it in `postprocess` - this is the correct way but will result in different outputs. I did not want to change expected slices so I denormalized input before passing it to VaeImageProcessor - this way we will match the expected slice from the previous implementation

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py [hunk 3]
```diff
@@ -436,17 +435,32 @@ def _encode_prompt(
...
+                feature_extractor_input = self.image_processor.postprocess(image, output_type="pil")
+            else:
+                feature_extractor_input = self.image_processor.numpy_to_pil(image)
+            safety_checker_input = self.feature_extractor(feature_extractor_input, return_tensors="pt").to(device)
+            image, has_nsfw_concept = self.safety_checker(
+                images=image, clip_input=safety_checker_input.pixel_values.to(dtype)
+            )
         return image, has_nsfw_concept
 
     def decode_latents(self, latents):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py:451]: the code change is a little bit confusing here but it's same as `decode_latents `in StableDiffusionPipeline only with the  deprecation message added

Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py:451]: Perfect! It's essentially a revert of what we had before

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 4]
```diff
@@ -738,27 +750,19 @@ def __call__(
...
 
+        if has_nsfw_concept is None:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:760]:
```suggestion
```

File: tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py [hunk 5]
```diff
@@ -96,45 +101,32 @@ def get_dummy_components(self):
...
         return components
 
-    def get_dummy_inputs(self, device, seed=0, input_image_type="pt", output_type="np"):
+    def get_dummy_inputs(self, device, seed=0):
         image = floats_tensor((1, 3, 32, 32), rng=random.Random(seed)).to(device)
         if str(device).startswith("mps"):
             generator = torch.manual_seed(seed)
         else:
             generator = torch.Generator(device=device).manual_seed(seed)
-
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py:105]: why delete this?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py (hunk 2 above)
Comment (patrickvonplaten) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py:129]: Ok for me

File: tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py [hunk 6]
```diff
@@ -217,36 +212,6 @@ def test_save_load_optional_components(self):
     def test_attention_slicing_forward_pass(self):
         return super().test_attention_slicing_forward_pass()
 
-    @skip_mps
-    def test_pt_np_pil_outputs_equivalent(self):
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py:221]: why do we delete the tests here?

**Review by patrickvonplaten — APPROVED** [review]
Great! This looks good to me!

**Review by pcuenca — APPROVED** [review]
LGTM!

File: src/diffusers/image_processor.py [hunk 7]
```diff
@@ -165,17 +172,39 @@ def preprocess(
...
         self,
-        image,
+        image: torch.FloatTensor,
         output_type: str = "pil",
+        do_denormalize: Optional[List[bool]] = None,
     ):
-        if isinstance(image, torch.Tensor) and output_type == "pt":
+        if not isinstance(image, torch.Tensor):
+            raise ValueError(
+                f"Input for postprocess is in incorrect format: {type(image)}.  we only support pytorch tensor"
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:181]:
```suggestion
                f"Input for postprocessing is in incorrect format: {type(image)}. We only support pytorch tensor"
```

File: src/diffusers/image_processor.py [hunk 8]
```diff
@@ -165,17 +172,39 @@ def preprocess(
...
+            output_type = "np"
+
+        if output_type == "latent":
+            return image
+
+        if do_denormalize is None:
+            do_denormalize = [self.config.do_normalize] * image.shape[0]
+
+        image = torch.stack(
+            [self.denormalize(image[i]) if do_denormalize[i] else image[i] for i in range(image.shape[0])]
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:198]: This is neat!

But would there be a need to denormalize images individually? I would assume if `do_denormalize` is True then ALL the images will be denormalized. Why individually? 

Nit: Should it be "unnormalize"?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 9]
```diff
@@ -443,17 +442,30 @@ def _encode_prompt(
...
+            safety_checker_input = self.feature_extractor(feature_extractor_input, return_tensors="pt").to(device)
+            image, has_nsfw_concept = self.safety_checker(
+                images=image, clip_input=safety_checker_input.pixel_values.to(dtype)
+            )
         return image, has_nsfw_concept
 
     def decode_latents(self, latents):
+        warnings.warn(
+            "The decode_latents method is deprecated and will be removed in a future version. Please"
+            " use VaeImageProcessor instead",
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:461]: Should we also specify the method of `VaeImagProcessor` users should use? I think that is more helpful and specific.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py (hunk 2 above)
Comment (sayakpaul) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py:129]: A comment above the line briefly explaining what you mentioned will be helpful for the future contributors.

Comment (sayakpaul) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py:129]: Applies here and elsewhere.

**Review by sayakpaul — APPROVED** [review]
Great work here. I think this is a really clean refactor. I would like to do one final pass on Monday if that's play and non-blocking.

File: src/diffusers/image_processor.py (hunk 8 above)
Comment (yiyixuxu) [reply src/diffusers/image_processor.py:198]: normally we should denormalize all images if we have normalized them in preprocessing (i.e. `if self.config.do_normalize is True`)
however, there is this edge case when the image is black due to `safety_checker` - in that case, we will pass a `do_denormalize` argument to allow skip the denormalize process only for these nsfw images.

Comment (sayakpaul) [reply src/diffusers/image_processor.py:198]: Superb! Thanks for explaining!

**Comment by sayakpaul** [comment]
Looks good to me. Took a final look.
